File: app/src/main/kotlin/tech/mmarca/openvitals/domain/preferences/ActivityRecordingPreferences.kt
package tech.mmarca.openvitals.domain.preferences

data class ActivityRecordingPreferences(
    val autoIdleEnabled: Boolean = DefaultAutoIdleEnabled,
    val autoIdleTimeoutSeconds: Int = DefaultAutoIdleTimeoutSeconds,
    val keepScreenOnDuringRecording: Boolean = DefaultKeepScreenOnDuringRecording,
    val requiredGpsAccuracyMeters: Int = DefaultRequiredGpsAccuracyMeters,
    val routeGapMeters: Int? = DefaultRouteGapMeters,
    val barometerClimbEnabled: Boolean = DefaultBarometerClimbEnabled,
    val recordingDistanceIntervalMeters: Int? = DefaultRecordingDistanceIntervalMeters,
    val recordingTimeIntervalMillis: Int = DefaultRecordingTimeIntervalMillis,
    val voiceAnnouncementsEnabled: Boolean = DefaultVoiceAnnouncementsEnabled,
    val voiceAnnouncementTimeIntervalMinutes: Int? = DefaultVoiceAnnouncementTimeIntervalMinutes,
    val voiceAnnouncementDistanceIntervalMeters: Int? = DefaultVoiceAnnouncementDistanceIntervalMeters,
    val voiceIdleAnnouncementsEnabled: Boolean = DefaultVoiceIdleAnnouncementsEnabled,
    val voiceLapAnnouncementsEnabled: Boolean = DefaultVoiceLapAnnouncementsEnabled,
) {
    fun normalized(): ActivityRecordingPreferences =
        copy(
            autoIdleTimeoutSeconds = autoIdleTimeoutSeconds.coerceIn(
                MinAutoIdleTimeoutSeconds,
                MaxAutoIdleTimeoutSeconds,
            ),
            requiredGpsAccuracyMeters = requiredGpsAccuracyMeters.closestAllowed(AllowedGpsAccuracyMeters),
            routeGapMeters = routeGapMeters?.closestAllowed(AllowedRouteGapMeters),
            recordingDistanceIntervalMeters = recordingDistanceIntervalMeters
                ?.closestAllowed(AllowedRecordingDistanceIntervalMeters),
            recordingTimeIntervalMillis = recordingTimeIntervalMillis
                .closestAllowed(AllowedRecordingTimeIntervalMillis),
            voiceAnnouncementTimeIntervalMinutes = voiceAnnouncementTimeIntervalMinutes
                ?.closestAllowed(AllowedVoiceAnnouncementTimeIntervalMinutes),
            voiceAnnouncementDistanceIntervalMeters = voiceAnnouncementDistanceIntervalMeters
                ?.closestAllowed(AllowedVoiceAnnouncementDistanceIntervalMeters),
        )

    companion object {
        const val DefaultAutoIdleEnabled = true
        const val DefaultAutoIdleTimeoutSeconds = 10
        const val MinAutoIdleTimeoutSeconds = 5
        const val MaxAutoIdleTimeoutSeconds = 60
        const val DefaultKeepScreenOnDuringRecording = false
        const val DefaultRequiredGpsAccuracyMeters = 30
        const val DefaultBarometerClimbEnabled = true
        val DefaultRecordingDistanceIntervalMeters: Int? = null
        const val DefaultRecordingTimeIntervalMillis = 500
        const val DefaultVoiceAnnouncementsEnabled = false
        val DefaultVoiceAnnouncementTimeIntervalMinutes: Int? = 5
        val DefaultVoiceAnnouncementDistanceIntervalMeters: Int? = 1_000
        const val DefaultVoiceIdleAnnouncementsEnabled = true
        const val DefaultVoiceLapAnnouncementsEnabled = true
        val DefaultRouteGapMeters: Int? = 200
        val AllowedGpsAccuracyMeters = listOf(10, 30, 50, 100)
        val AllowedRouteGapMeters = listOf(100, 200, 500)
        val AllowedRecordingDistanceIntervalMeters = listOf(5, 10, 25, 50)
        val AllowedRecordingTimeIntervalMillis = listOf(500, 1_000, 5_000, 10_000)
        val AllowedVoiceAnnouncementTimeIntervalMinutes = listOf(1, 5, 10)
        val AllowedVoiceAnnouncementDistanceIntervalMeters = listOf(500, 1_000, 5_000)
    }
}

private fun Int.closestAllowed(allowedValues: List<Int>): Int =
    allowedValues.minBy { kotlin.math.abs(it - this) }

File: lib/domain/preferences/activity_recording_preferences.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'activity_recording_preferences.freezed.dart';

/// Port of the Kotlin `ActivityRecordingPreferences` — GPS/activity recording
/// tuning used by the (deferred) activity-recording feature. Persisted through
/// `PreferencesRepository`.
@freezed
abstract class ActivityRecordingPreferences with _$ActivityRecordingPreferences {
  const ActivityRecordingPreferences._();

  const factory ActivityRecordingPreferences({
    @Default(ActivityRecordingPreferences.defaultAutoIdleEnabled)
    bool autoIdleEnabled,
    @Default(ActivityRecordingPreferences.defaultAutoIdleTimeoutSeconds)
    int autoIdleTimeoutSeconds,
    @Default(ActivityRecordingPreferences.defaultKeepScreenOnDuringRecording)
    bool keepScreenOnDuringRecording,
    @Default(ActivityRecordingPreferences.defaultRequiredGpsAccuracyMeters)
    int requiredGpsAccuracyMeters,
    @Default(ActivityRecordingPreferences.defaultRouteGapMeters)
    int? routeGapMeters,
    @Default(ActivityRecordingPreferences.defaultBarometerClimbEnabled)
    bool barometerClimbEnabled,
    int? recordingDistanceIntervalMeters,
    @Default(ActivityRecordingPreferences.defaultRecordingTimeIntervalMillis)
    int recordingTimeIntervalMillis,
    @Default(ActivityRecordingPreferences.defaultVoiceAnnouncementsEnabled)
    bool voiceAnnouncementsEnabled,
    @Default(
      ActivityRecordingPreferences.defaultVoiceAnnouncementTimeIntervalMinutes,
    )
    int? voiceAnnouncementTimeIntervalMinutes,
    @Default(
      ActivityRecordingPreferences
          .defaultVoiceAnnouncementDistanceIntervalMeters,
    )
    int? voiceAnnouncementDistanceIntervalMeters,
    @Default(ActivityRecordingPreferences.defaultVoiceIdleAnnouncementsEnabled)
    bool voiceIdleAnnouncementsEnabled,
    @Default(ActivityRecordingPreferences.defaultVoiceLapAnnouncementsEnabled)
    bool voiceLapAnnouncementsEnabled,
    @Default(ActivityRecordingPreferences.defaultRestTimerBellEnabled)
    bool restTimerBellEnabled,
  }) = _ActivityRecordingPreferences;

  ActivityRecordingPreferences normalized() => copyWith(
        autoIdleTimeoutSeconds: autoIdleTimeoutSeconds
            .clamp(minAutoIdleTimeoutSeconds, maxAutoIdleTimeoutSeconds)
            .toInt(),
        requiredGpsAccuracyMeters: _closestAllowed(
          requiredGpsAccuracyMeters,
          allowedGpsAccuracyMeters,
        ),
        routeGapMeters: routeGapMeters == null
            ? null
            : _closestAllowed(routeGapMeters!, allowedRouteGapMeters),
        recordingDistanceIntervalMeters: recordingDistanceIntervalMeters == null
            ? null
            : _closestAllowed(
                recordingDistanceIntervalMeters!,
                allowedRecordingDistanceIntervalMeters,
              ),
        recordingTimeIntervalMillis: _closestAllowed(
          recordingTimeIntervalMillis,
          allowedRecordingTimeIntervalMillis,
        ),
        voiceAnnouncementTimeIntervalMinutes:
            voiceAnnouncementTimeIntervalMinutes == null
                ? null
                : _closestAllowed(
                    voiceAnnouncementTimeIntervalMinutes!,
                    allowedVoiceAnnouncementTimeIntervalMinutes,
                  ),
        voiceAnnouncementDistanceIntervalMeters:
            voiceAnnouncementDistanceIntervalMeters == null
                ? null
                : _closestAllowed(
                    voiceAnnouncementDistanceIntervalMeters!,
                    allowedVoiceAnnouncementDistanceIntervalMeters,
                  ),
      );

  static int _closestAllowed(int value, List<int> allowedValues) {
    var best = allowedValues.first;
    var bestDistance = (best - value).abs();
    for (final candidate in allowedValues.skip(1)) {
      final distance = (candidate - value).abs();
      if (distance < bestDistance) {
        best = candidate;
        bestDistance = distance;
      }
    }
    return best;
  }

  static const bool defaultAutoIdleEnabled = true;
  static const int defaultAutoIdleTimeoutSeconds = 10;
  static const int minAutoIdleTimeoutSeconds = 5;
  static const int maxAutoIdleTimeoutSeconds = 60;
  static const bool defaultKeepScreenOnDuringRecording = false;
  static const int defaultRequiredGpsAccuracyMeters = 30;
  static const bool defaultBarometerClimbEnabled = true;
  static const int? defaultRecordingDistanceIntervalMeters = null;
  static const int defaultRecordingTimeIntervalMillis = 500;
  static const bool defaultVoiceAnnouncementsEnabled = false;
  static const int defaultVoiceAnnouncementTimeIntervalMinutes = 5;
  static const int defaultVoiceAnnouncementDistanceIntervalMeters = 1000;
  static const bool defaultVoiceIdleAnnouncementsEnabled = true;
  static const bool defaultVoiceLapAnnouncementsEnabled = true;
  static const bool defaultRestTimerBellEnabled = true;
  static const int defaultRouteGapMeters = 200;
  static const List<int> allowedGpsAccuracyMeters = [10, 30, 50, 100];
  static const List<int> allowedRouteGapMeters = [100, 200, 500];
  static const List<int> allowedRecordingDistanceIntervalMeters = [5, 10, 25, 50];
  static const List<int> allowedRecordingTimeIntervalMillis = [
    500,
    1000,
    5000,
    10000,
  ];
  static const List<int> allowedVoiceAnnouncementTimeIntervalMinutes = [1, 5, 10];
  static const List<int> allowedVoiceAnnouncementDistanceIntervalMeters = [
    500,
    1000,
    5000,
  ];
}
